src/main/java/com/aws/greengrass/shadowmanager/ShadowManagerDAOImpl.java

```diff
@@ -352,6 +355,16 @@ public boolean insertSyncInfoIfNotExists(SyncInformation request) {
 
     }
 
+    @Override
+    public void waitForDBOperationsToFinish() {
+        if (!dbWriteOperations.tryAcquire(maxPermits)) {
```

Comment:
Why call this and then if it fails to acquire we wait call the blocking version?. We want to basically ensure we grab all the permits right?

```java
logger.atDebug().log("Waiting for the DB write operations to finish");
dbWriteOperations.acquireUninterruptibly(maxPermits);
```

Also this method partially works, we are acquiring all the permits and immediately releasing them there is no guarantee that another thread can grab a permit right after we release them that might lead to corruption. (I know you are calling this on the release method but given there are not tests for this someone in the future might just see this and remove it given they might not have the context.


Is there a way we could reject all write requests after this is called?  Maybe reducing the permits to 0?

Reply:
Calling this code from the db close method now.